src/org/github/jamm/MemoryLayoutSpecification.java

```diff
@@ -1,144 +1,126 @@
+
+    /**
+     * Returns the object alignment (padding) in bytes.
+     * <p>The alignment is always a power of 2.</p>
+     *
+     * @return the object alignment in bytes.
+     */
+    int getObjectAlignment();
+
+    /**
+     * Returns the size of the reference to java objects (also called <i>oops</i> for <i>ordinary object pointers</i>)
+     *
+     * @return the java object reference size
+     */
+    int getReferenceSize();
 
     public static MemoryLayoutSpecification getEffectiveMemoryLayoutSpecification() {
 
-        final String dataModel = System.getProperty("sun.arch.data.model");
-        if ("32".equals(dataModel)) {
+        int objectAlignment = VM.getObjectAlignmentInBytes();
+
+        if (VM.is32Bits()) {
             // Running with 32-bit data model
             return new MemoryLayoutSpecification() {
+
+                @Override
                 public int getArrayHeaderSize() {
-                    return 12;
+                    return 12; // object header (8 bytes) + array length (4 bytes)
                 }
 
+                @Override
                 public int getObjectHeaderSize() {
-                    return 8;
+                    return 8; // mark word (4 bytes) + class word (4 bytes)
                 }
 
-                public int getObjectPadding() {
-                    return 8;
+                @Override
+                public int getObjectAlignment() {
+                    return objectAlignment;
                 }
 
+                @Override
                 public int getReferenceSize() {
-                    return 4;
-                }
-
-                public int getSuperclassFieldPadding() {
-                    return 4;
+                    return 4; // reference size for 32 bit
                 }
             };
         }
 
-        boolean modernJvm = true;
-
-        final String strSpecVersion = System.getProperty("java.specification.version");
-        final boolean hasDot = strSpecVersion.indexOf('.') != -1;
-        if (hasDot) {
-            if ("1".equals(strSpecVersion.substring(0, strSpecVersion.indexOf('.')))) {
-                // Java 1.6, 1.7, 1.8
-                final String strVmVersion = System.getProperty("java.vm.version");
-                if (strVmVersion.startsWith("openj9"))
-                {
-                    modernJvm = true;
-                }
-                else
-                {
-                    final int vmVersion = Integer.parseInt(strVmVersion.substring(0, strVmVersion.indexOf('.')));
-                    modernJvm = vmVersion >= 17;
-                }
-            }
-        }
-
-        final int alignment = getAlignment();
-        if (modernJvm) {
-
-            long maxMemory = 0;
-            for (MemoryPoolMXBean mp : ManagementFactory.getMemoryPoolMXBeans()) {
-                maxMemory += mp.getUsage().getMax();
-            }
+        if (VM.useCompressedOops()) {
 
-            if (maxMemory < 30L * 1024 * 1024 * 1024) {
```

Comment:
Compressed references can be turned on or off depending on the Heap size (https://shipilev.net/jvm/anatomy-quarks/23-compressed-references/) but `VM.useCompressedOops()` will return the correct value that the JVM making this code useless (it was also wrong for another scenario).

Reply:
I was thinking of the case if someone sets explicitly `-XX:+UseCompressedOops` and something big for  `-Xmx` like 40GB because they didn't know/made a mistake. Like in theory they shouldn't but someone might have missed the 32GB limitation. I am wondering whether the JVM will be just ignoring `-XX:+UseCompressedOops` or it will switch internally and report `-XX:-UseCompressedOops`, it seems it is switching it to false, if I am looking at the right place:
[If I look at the right place it seems that is what ](https://github.com/AdoptOpenJDK/openjdk-jdk8u/blob/master/hotspot/src/share/vm/runtime/arguments.cpp#L1535-L1538)
Also:
https://github.com/openjdk/jdk/blob/master/src/hotspot/share/runtime/arguments.cpp#L1489-L1508

Reply:
I ran some tests with different heap and the result returned by JMX where always the correct ones. So we should be good.

Reply:
This code is guessing a behavior from the JVM when the JVM can actually give you with the correct information in which you are interested are compressed OOPs enable, yes or no. I ran some check and the JVM always return the correct information. People tend to copy the approach of others that does not mean that it is the right one.